subprojects/composite-builds/src/integTest/groovy/org/gradle/integtests/composite/CompositeBuildContinueOnFailureIntegrationTest.groovy

```diff
@@ -159,4 +142,14 @@ class CompositeBuildContinueOnFailureIntegrationTest extends AbstractCompositeBu
         executed ":buildB:jar", ":buildB:checkContinueFlag"
         outputContains("continueOnFailure = true")
     }
+
+    // With failing tasks and parallel output, we cannot assert that the task was executed only once.
+    // This is because our OutputScrapingExecutionResult doesn't handle the output for interleaved tasks
+    // TODO:DAZ Using build operations instead of output scraping would help.
+    protected void executed(String... tasks) {
+        for (String task : tasks) {
+            assert task in executedTasks
+        }
+    }
+
```

Comment:
When an build artifact has an 'undeclared' dependency that is substituted, then this task may be added later, while other tasks are executing. An example is a `compileOnly` dependency.  I guess we should add a test where an included build has both a `compile` and `compileOnly` dependency on different outputs from another included build: this will validate the ability to kick off a second build execution for the late-discovered dependency.

In the long term, we should try to discover as many of these task dependencies up-front as possible, to avoid executing an included build multiple times.